plugins/CircuitViewer/io/CircuitLoader.cpp

```diff
@@ -584,72 +672,75 @@ class Impl
+
+            const auto morphologies =
+                circuit.loadMorphologies(batch,
+                                         brain::Circuit::Coordinates::global);
+
+#pragma omp parallel for
+            for (uint64_t j = 0; j < morphologies.size(); ++j)
             {
-                ++current;
+                const auto morphologyIndex = index + j;
+                auto materialFunc = [&](const brain::neuron::SectionType type) {
+                    if (_properties.useSimulationModel)
+                        return size_t{0};
+                    else
+                        return _getMaterialId(_properties.colorScheme,
+                                              morphologyIndex, type,
+                                              perCellMaterialIds);
+                };
+
+                const auto& morphology = morphologies[j];
+
+                auto data =
+                    morphLoader.processMorphology(*morphology, morphologyIndex,
+                                                  materialFunc, reportMapping,
+                                                  _morphologyParams);
+#pragma omp critical
+                data.dump(model);
 
+                ++current;
+                // Since throwing from inside an parallel for is not a good
+                // idea. If the loading gets interrupted the exception will
+                // be caught here and rethrown later after exiting the loop.
                 try
                 {
                     callback.updateProgress(message.str(),
-                                            current / static_cast<float>(
-                                                          uris.size()));
-                    const auto& uri = uris[morphologyIndex];
-
-                    auto materialFunc =
-                        [&](const brain::neuron::SectionType type) {
-                            if (_properties.useSimulationModel)
-                                return size_t{0};
-                            else
-                                return _getMaterialId(_properties.colorScheme,
-                                                      morphologyIndex, type,
-                                                      perCellMaterialIds);
-                        };
-
-                    ModelData modelData;
-                    if (!morphLoader._importMorphology(
-                            uri, morphologyIndex, materialFunc,
-                            transformations[morphologyIndex], reportMapping,
-                            modelData, _morphologyParams))
-#pragma omp atomic
-                        ++loadingFailures;
-#pragma omp critical
-                    modelData.addTo(model);
+                                            current / float(gids.size()));
                 }
                 catch (...)
                 {
                     cancelException = std::current_exception();
-                    morphologyIndex = uris.size();
```

Comment:
The loading then continues until `uris.size()`, even though it was cancelled?!

Reply:
There was something missing in the outer loop to stop the iteration if the cancel exception is raised.